### PR TITLE
Implement resolve for facet and concat

### PR DESCRIPTION
**Pull request by domoritz**
screen shot 2017-05-24 at 16 38 31

screen shot 2017-05-24 at 19 25 19

### CODE REVIEW

**Comment by domoritz** [comment]
Done and ready.

File: src/compile/scale/domain.ts [hunk 1]
```diff
@@ -235,15 +228,7 @@ function normalizeDomain(domain: UnionableDomain): (any[] | VgDataRef)[] {
...
-    return domain.fields.map(d => {
-      if (util.isArray(d)) {
-        return d;
-      }
-      return {
-        data: d.data,
-        field: d.field
-      };
-    });
+    return domain.fields;
```
Comment (kanitw) [inline src/compile/scale/domain.ts]: Can you explain why is this part changing?

File: src/compile/scale/parse.ts [hunk 2]
```diff
@@ -65,3 +66,27 @@ export function parseScale(model: UnitModel, channel: Channel) {
...
 }
+
+export function moveSharedScaleUp(model: Model, scaleComponent: Dict<ScaleComponent>, child: Model, channel: Channel) {
+  // Check whether the scales are actually compatible, e.g. use the same sort or throw error
+
+  const childScale = child.component.scales[channel];
+  let modelScale = scaleComponent[channel];
+
+  if (modelScale) {
+    modelScale.domain = unionDomains(modelScale.domain, childScale.domain);
```
Comment (kanitw) [inline src/compile/scale/parse.ts]: Add `// TODO: merge properties if they are not conflicting`

(Beyond the scope of this PR.)

File: src/compile/scale/domain.ts [hunk 3]
```diff
@@ -275,10 +253,12 @@ export function unionDomains(domain1: VgDomain, domain2: VgDomain): VgDomain {
     });
 
     if (util.unique(allData, x => x).length === 1 && allData[0] !== null) {
-      return {
+      // create a union domain of different fields with a single data source
+      const domain: FieldRefUnionDomain = {
```
Comment (kanitw) [inline src/compile/scale/domain.ts]: not sure why returning right away doesn't work anymore..

File: src/compile/concat.ts [hunk 4]
```diff
@@ -77,6 +92,16 @@ export class ConcatModel extends Model {
 
     for (const child of this.children) {
       child.parseLegend();
+
+      keys(child.component.legends).forEach((channel: ScaleChannel) => {
+        if (this.resolve[channel].legend === 'shared') {
+          // just use the first legend definition for each channel
```
Comment (kanitw) [inline src/compile/concat.ts]: Add `// TODO: merge properties if they are not conflicting`

(This is still wrong, but the fix is beyond the scope of this PR.)

Comment (kanitw) [inline src/compile/concat.ts]: Isn't this logic shared by other models?  Extract?

Comment (domoritz) [inline src/compile/concat.ts]: Good catch. Done.

File: src/compile/scale/parse.ts (hunk 2 above)
Comment (domoritz) [inline src/compile/scale/parse.ts]: done

**Pull request merged**